### DISCUSSION TITLE
ASP3カーネル (for mros2) ビルド時の警告メッセージについて

### DISCUSSION THREAD

**Discussion by takasehideki** [toppers/users-forum#53] [Q&A] [answered]
@hiroakitakada @hiro22022 @aztk Cc: @yurie @ykominami 
ASP3カーネルを弊mros2向けに魔改造した sample1 のビルド時に，警告メッセージが表示されます．動作上は問題ありません．
亜種のものですしTECSとcfgの双方に関わるため，どこに投稿すべきか悩みまして，ここに投下させていただきます．警告メッセージの抑止策について，なにかお気づきの点やヒントなどご存知でしたらご教示ください．

# 試行環境

## リポジトリ・ボード等の情報

- リポジトリ： https://github.com/mROS-base/asp3-f767zi
  - TOPPERS/ASP3カーネル Release 3.5.0
  - tecsgen version 1.7.0
  - https://github.com/yurie/asp3-f767zi を魔改造
- ボード： STM32 Nucleo-F767ZI

## ホスト環境

- Ubuntu 20.04.3 LTS
- gcc-arm-none-eabi, ruby は apt package

```
$ uname -a 
Linux vaioubuntu20 5.4.0-91-generic #102-Ubuntu SMP Fri Nov 5 16:31:28 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

$ lsb_release -a 
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.3 LTS
Release:	20.04
Codename:	focal

$ arm-none-eabi-gcc -v 
Using built-in specs.
COLLECT_GCC=arm-none-eabi-gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/arm-none-eabi/9.2.1/lto-wrapper
Target: arm-none-eabi
Configured with: ../configure --build=x86_64-linux-gnu --prefix=/usr --includedir='/usr/lib/include' --mandir='/usr/lib/share/man' --infodir='/usr/lib/share/info' --sysconfdir=/etc --localstatedir=/var --disable-silent-rules --libdir='/usr/lib/lib/x86_64-linux-gnu' --libexecdir='/usr/lib/lib/x86_64-linux-gnu' --disable-maintainer-mode --disable-dependency-tracking --mandir=/usr/share/man --enable-languages=c,c++,lto --enable-multilib --disable-decimal-float --disable-libffi --disable-libgomp --disable-libmudflap --disable-libquadmath --disable-libssp --disable-libstdcxx-pch --disable-nls --disable-shared --disable-threads --enable-tls --build=x86_64-linux-gnu --target=arm-none-eabi --with-system-zlib --with-gnu-as --with-gnu-ld --with-pkgversion=15:9-2019-q4-0ubuntu1 --without-included-gettext --prefix=/usr/lib --infodir=/usr/share/doc/gcc-arm-none-eabi/info --htmldir=/usr/share/doc/gcc-arm-none-eabi/html --pdfdir=/usr/share/doc/gcc-arm-none-eabi/pdf --bindir=/usr/bin --libexecdir=/usr/lib --libdir=/usr/lib --disable-libstdc++-v3 --host=x86_64-linux-gnu --with-headers=no --without-newlib --with-multilib-list=rmprofile CFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-Gl9kT9/gcc-arm-none-eabi-9-2019-q4=. -fstack-protector-strong' CPPFLAGS='-Wdate-time -D_FORTIFY_SOURCE=2' CXXFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-Gl9kT9/gcc-arm-none-eabi-9-2019-q4=. -fstack-protector-strong' FCFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-Gl9kT9/gcc-arm-none-eabi-9-2019-q4=. -fstack-protector-strong' FFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-Gl9kT9/gcc-arm-none-eabi-9-2019-q4=. -fstack-protector-strong' GCJFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-Gl9kT9/gcc-arm-none-eabi-9-2019-q4=. -fstack-protector-strong' LDFLAGS='-Wl,-Bsymbolic-functions -Wl,-z,relro' OBJCFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-Gl9kT9/gcc-arm-none-eabi-9-2019-q4=. -fstack-protector-strong' OBJCXXFLAGS='-g -O2 -fdebug-prefix-map=/build/gcc-arm-none-eabi-Gl9kT9/gcc-arm-none-eabi-9-2019-q4=. -fstack-protector-strong' INHIBIT_LIBC_CFLAGS=-DUSE_TM_CLONE_REGISTRY=0 AR_FOR_TARGET=arm-none-eabi-ar AS_FOR_TARGET=arm-none-eabi-as LD_FOR_TARGET=arm-none-eabi-ld NM_FOR_TARGET=arm-none-eabi-nm OBJDUMP_FOR_TARGET=arm-none-eabi-objdump RANLIB_FOR_TARGET=arm-none-eabi-ranlib READELF_FOR_TARGET=arm-none-eabi-readelf STRIP_FOR_TARGET=arm-none-eabi-strip
Thread model: single
gcc version 9.2.1 20191025 (release) [ARM/arm-9-branch revision 277599] (15:9-2019-q4-0ubuntu1) 

$ ruby -v
ruby 2.7.0p0 (2019-12-25 revision 647ee6f091) [x86_64-linux-gnu]
```

# 試行方法

https://github.com/mROS-base/asp3-f767zi#how-to-build-sample1-app-without-mros-feature

# ビルドログと思索

```
$ cd OBJ
$ make realclean
$ make 2>&1 | tee make.log
```

[make.log](https://github.com/toppers/users-forum/files/7750120/make.log)

## TECSに関わるもの？

make.log 3行目から220行目

```
/home/takase/ros/mROS2/github/mros2-asp3-f767zi/asp3-f767zi/tecsgen/tecslib/core/bnf.tab.rb:5156: warning: Capturing the given block using Proc.new is deprecated; use `&block` instead
/home/takase/ros/mROS2/github/mros2-asp3-f767zi/asp3-f767zi/tecsgen/tecslib/core/componentobj.rb:4521: warning: Capturing the given block using Proc.new is deprecated; use `&block` instead
```

前者は， GitHub上だと https://github.com/mROS-base/asp3-f767zi/blob/mros2/tecsgen/tecslib/core/bnf.tab.rb#L5156
TOPPERS Trac上だと https://dev.toppers.jp/trac/tecs/browser/toppers/tecsgen/trunk/tecsgen/tecslib/core/bnf.tab.rb#L5124
に該当するところと思われる

後者は，GitHub上だと https://github.com/mROS-base/asp3-f767zi/blob/mros2/tecsgen/tecslib/core/componentobj.rb#L4521
TOPPERS Trac上だと https://dev.toppers.jp/trac/tecs/browser/toppers/tecsgen/trunk/tecsgen/tecslib/core/componentobj.rb 4613行目？（ファイルサイズが大きくプレビューできず）
 
いずれもTrac最新ではコメントアウトされている．
これに従って両者をコメントアウトしたら，別のエラーが発生してしまう．
警告メッセージを抑止するには最新版を持ってきたら良さそうだが，どうやったらよいか？？

## cfgに関わるもの？

make.log 222-225行目

```
ruby ../cfg/cfg.rb --pass 1 --kernel asp -I. -I../include  -I../target/nucleo_f767zi_gcc -I../target/nucleo_f767zi_gcc/stm32fcube -I../arch/arm_m_gcc/stm32f7xx_stm32cube/ -I../../STM32CubeF7/Drivers/STM32F7xx_HAL_Driver/Inc -I../../STM32CubeF7/Drivers/CMSIS/Device/ST/STM32F7xx/Include -I../../STM32CubeF7/Drivers/CMSIS/Include -I../arch/arm_m_gcc/common -I../arch/gcc -I.. -I../sample -I./gen -I../tecs_kernel --api-table ../kernel/kernel_api.def --symval-table ../kernel/kernel_sym.def  --symval-table ../arch/arm_m_gcc/common/core_sym.def \
					-M objs/cfg1_out_c.d ../target/nucleo_f767zi_gcc/target_kernel.cfg ../sample/sample1.cfg
/home/takase/ros/mROS2/github/mros2-asp3-f767zi/asp3-f767zi/cfg/pass1.rb:178: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/usr/lib/ruby/2.7.0/csv.rb:635: warning: The called method `open' is defined here
```

GitHub上だと https://github.com/mROS-base/asp3-f767zi/blob/mros2/cfg/pass1.rb#L178-L179
TOPPERS Trac上だと https://dev.toppers.jp/trac/ruby_cfg/browser/trunk/cfg/pass1.rb#L180
に該当するところと思われる

これはまだ特に対処されていないように思われるが，どうするとよいか？？

**Comment by h7ga40** [comment]
cfgに関しては、TOPPERS Tracのリビジョン191で対処されていると思います。
「pass1.rbを，rubyの新しいバージョンでエラーにならないように修正。」のコメントがあります。

**Reply by takasehideki** [reply]
@h7ga40 ありがとうございます！
よく見たら `{` `}` が取れているのを完全に見落としておったようです,,, 
https://dev.toppers.jp/trac/ruby_cfg/changeset/191/

こちら軽微な修正ですし他への影響も無いことが確認できたので，ここだけ弊リポジトリに取り込むことにしたいと思います．

**Comment by ykominami** [comment]
「TECSに関わるもの」に関しては、私が1年以上前にTOPPERSプロジェクト内部のTECS WGのTracにチケットとして登録(1186)しており、TOPPERS内部のリポジトリ上ではfix済みです。
チケットにはwarningが出なくなることを確認したソースコード変更のパッチを添付してあります。

Proc.newの行をコメントアウトするだけではなく、Proc.newを呼び出したメソッドの引数として、ブロック引数&prを追加する必要があります。
また、ここでは指摘されていませんが、tecslib/core/generate.rbの940行付近のメソッドtravers_all_signatureにも同様の対応が必要です。

現状は、内部ではfix済みですが、最近はTECS WGからのリリースがなかったため、fix済みのバージョンのtecsgenは公開されていません。

質問されたwarningが発生する原因は、Ruby2.7からProc.newの振る舞いが変更されたためです。
詳しくは、例えば以下のRubyの公式のドキュメントを参照して下さい。
https://docs.ruby-lang.org/ja/latest/doc/news=2f2_7_0.html

Ruby2.7ではwarningの発生だけですが、Ruby3.0以降では例外が発生してしまいます。

また、tecslib/core/bnf.tab.rbは、YACCに似たRubyのパーサジェネレータの一つであるRaccが自動生成したファイルです。
warningを発生させない対応をする場合は、このファイルではなく、このファイルの生成元となるtecslib/core/bnf.y.rbの方を修正してください。

「cfgに関わるもの？」に関しても、TOPPERSプロジェクト内部でのリポジトリでは対応されているようです。
ただし、直近に公開されたASP3 RX簡易パッケージに同梱されているcfgには反映されていません。
https://www.toppers.jp/download.cgi/asp3_rx_ccrx-20211104.tar.gz

これもRuby 2.7からのメソッドのキーワード引数の振る舞いが変更されたことによります。
対応方法は上記URLの公式のドキュメントに記載されています。

warningを出さないようにする方法ですが、以下の方法が考えられます。

1. 用いるRubyのバージョンを2.6以前にする
実行するrubyのバージョンにこだわりがなく、Rubyからのwarningに悩まず他のことに注力したい場合、これが最も容易な方法です。
2. （私が今まで書いてきたことを参考にして）独自にRubyスクリプトを修正
Ruby2.7なら、warningですみますが、Ruby3.0以降ではエラーとか、例外発生の可能性もあるため、どのバージョンのRubyでも実行できるようにしたい場合は、修正する必要があります。

**Comment by hiro22022** [comment, marked as answer]
高瀬君、小南さん

大山です

小南さんが書いてくださっているとおりです。
Ruby 2.6 以前なら警告が出ません。
Ruby 2.7 では、とりあえず警告を無視して問題ありません。
Ruby 3.0 では例外が発生し動作しません。

TECS ジェネレータの修正済みの版はあるのですが、
リリースできておりません。
先日、少し大きな修正を入れて、不具合が出ているので、
それがおわったら、リリースすることにいたします。
年内目標ですすめます。

以上です

2021年12月22日(水) 8:23 ykominami ***@***.***>:

**Comment by takasehideki** [comment]
@hiro22022 @ykominami 情報ありがとうございます．Trac/SVNではすでに対処済みとのこと，承知しました．

基本的／私的には，mros2ユーザに対してわざわざRubyのダウングレードを強いることはしたくないです．いっぽうでこちらの一部取り込みもできそうではありますが，他への影響が心配です．Ruby 2.7.0では警告で済んでいますし，直近にリリース予定となる tecsgen version 1.8.0 を取り込むのが良さそうだと感じました．リリースが完了しましたらお知らせいただけますと幸いです．

**Comment by yurie** [comment]
25日あたりにRuby3.1がリリースされるようなので、このタイミングでしたら一度そちらでもビルドできることを確認してリリースすると良いのではないでしょうか。

**Comment by takasehideki** [comment]
@hiro22022 @h7ga40 @yurie @ykominami 改めまして，皆さま情報やアドバイスをありがとうございました．
たいへん遅くなりましたが，TECS 1.8.0へのアップデートを試行し，当該の警告メッセージを抑止することができました．

- ASP3の対象プロジェクトのディレクトリ内にある tecsgen/ をいったん削除
- tecsgen-1.8.0.tgz をダウンロード＆解凍して配置 https://www.toppers.jp/tecs.html
- (`git add tecsgen/` してGit/GitHub管理下に追加)

という感じで進めましたが，お作法としてなにかよろしくないものでしたらご教示ください．

---
@yurie よろしくどうぞ〜（いちどミスってスミマセン,,, 
https://github.com/yurie/asp3-f767zi/pull/6
